// @lexical/html/LexicalHtml.mjs
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as modDev from './LexicalHtml.dev.mjs';
import * as modProd from './LexicalHtml.prod.mjs';
const mod = process.env.NODE_ENV !== 'production' ? modDev : modProd;
export const $generateDOMFromNodes = mod.$generateDOMFromNodes;
export const $generateDOMFromRoot = mod.$generateDOMFromRoot;
export const $generateHtmlFromNodes = mod.$generateHtmlFromNodes;
export const $generateNodesFromDOM = mod.$generateNodesFromDOM;
export const $getRenderContextValue = mod.$getRenderContextValue;
export const $withRenderContext = mod.$withRenderContext;
export const DOMRenderExtension = mod.DOMRenderExtension;
export const RenderContextExport = mod.RenderContextExport;
export const RenderContextRoot = mod.RenderContextRoot;
export const contextUpdater = mod.contextUpdater;
export const contextValue = mod.contextValue;
export const createRenderState = mod.createRenderState;
export const domOverride = mod.domOverride;
